### PR TITLE
Mve3 channel

### PR DESCRIPTION
**Pull request by HelloThere-ImMat**

### CODE REVIEW

File: dependencies/Channel.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+/*   Updated: 2023/12/04 14:40:14 by rbroque          ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include <algorithm>
+#include <map>
+
+#include "Client.hpp"
+
+// DEBUGING
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: check if it's required

File: dependencies/Channel.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+class Channel {
+   public:
+	Channel(std::string name, Client *const client);
+	~Channel();
+	void								   addNewUser(Client *const client);
+	void								   removeUser(Client *const client);
+	const std::string					   getUserList();
+	const std::string					  &getName() const;
+	std::map<std::string, SpecifiedClient> userMap;
+	std::string							   topic;
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: put in private (topic & userMap)

File: dependencies/Channel.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+   private:
+	std::string _name;
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: const

File: dependencies/Client.hpp [hunk 4]
```diff
@@ -26,7 +26,7 @@
 
 class Client {
    public:
-	Client();
+	Client(const Client &ref);
```
Comment (romann-broque) [inline dependencies/Client.hpp]: remove this constructor

File: dependencies/ClientManager.hpp [hunk 5]
```diff
@@ -15,6 +15,8 @@
 #include <map>
 
 #include "Client.hpp"
+// debug
+#include <iostream>
```
Comment (romann-broque) [inline dependencies/ClientManager.hpp:19]: check if it's required (else -> remove)

File: dependencies/Server.hpp [hunk 6]
```diff
@@ -55,10 +56,18 @@
 
 #define RPL_WELCOME "001 <client> :Welcome to the <networkname> Network, <nick>"
 
-// Message
+// Specified Message
+
+#define JOIN_MESSAGE   "JOIN :"
```
Comment (romann-broque) [inline dependencies/Server.hpp]: -> JOIN_PREFIX & PART_PREFIX

File: dependencies/Server.hpp [hunk 7]
```diff
@@ -100,13 +109,16 @@ class Server {
...
+	std::map<std::string, Channel *>	   _channels;
 	std::map<std::string, CommandFunction> _cmdMap;
 	std::string							   _password;
 	ClientManager						   _clientMap;
 	// Private Methods
 	void printLog(const std::string &logMessage) const;
 	//    Send Methods
 	void sendMessage(const std::string &message, const int clientFd) const;
+	void sendPrivateMessage(const std::string &message, Client *const sender,
+		Client *const receiver) const;
```
Comment (romann-broque) [inline dependencies/Server.hpp]: add const to sender & receiver

File: dependencies/Server.hpp [hunk 8]
```diff
@@ -130,9 +142,14 @@ class Server {
...
 	void nick(const std::vector<std::string> &cmd, Client *const client);
 	void ping(const std::vector<std::string> &cmd, Client *const client);
+	void join(const std::vector<std::string> &cmd, Client *const client);
+	void privmsg(const std::vector<std::string> &cmd, Client *const client);
+	void part(const std::vector<std::string> &cmd, Client *const client);
 	void error(const std::string &message, Client *const client);
 	// CMD_UTILS
 	bool isNicknameAlreadyUsed(const std::string &nickname);
+	void sendJoinMessage(
+		Channel *channel, Client *const client, std::string channelName);
```
Comment (romann-broque) [inline dependencies/Server.hpp]: void sendJoinMessage(
		Channel *const channel, const Client *const client, const std::string &channelName);

File: srcs/Channel.cpp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+/*   Created: 2023/11/29 13:18:12 by mat               #+#    #+#             */
+/*   Updated: 2023/12/04 14:33:36 by rbroque          ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "Channel.hpp"
+
+// Static
+
+std::string getSpecifiedNick(SpecifiedClient &spClient) {
```
Comment (romann-broque) [inline srcs/Channel.cpp]: const SpecifiedClient &spClient

File: srcs/Channel.cpp [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	if (spClient.isOp)
+		return ("@" + spClient.client->getNickname());
```
Comment (romann-broque) [inline srcs/Channel.cpp]: "@" -> define OP_PREFIX

File: srcs/Channel.cpp [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+	return (spClient.client->getNickname());
+}
+
+// Methods
+
+Channel::Channel(std::string name, Client *const client) {
```
Comment (romann-broque) [inline srcs/Channel.cpp]: `Channel::Channel(const std::string &name, const Client *const client): _name(name)`

File: srcs/Channel.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+	SpecifiedClient spClient = {.client = client, .isOp = true};
+
+	_name = name;
```
Comment (romann-broque) [inline srcs/Channel.cpp]: remove `_name = name;`

File: srcs/Channel.cpp [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	userMap[client->getNickname()] = spClient;
+	std::cout << "Successfully created a Channel" << std::endl;
```
Comment (romann-broque) [inline srcs/Channel.cpp]: remove debug print

File: dependencies/Channel.hpp [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+#include <algorithm>
+#include <map>
+
+#include "Client.hpp"
+
+// DEBUGING
+#include <iostream>
+
+typedef struct s_specifiedClient {
+	Client *client;
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: const Client *const client

File: srcs/Channel.cpp [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+Channel::~Channel() { userMap.clear(); }
+
+void Channel::addNewUser(Client *const client) {
```
Comment (romann-broque) [inline srcs/Channel.cpp]: const Client *const client

File: srcs/Channel.cpp [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+	SpecifiedClient spClient = {.client = client, .isOp = false};
```
Comment (romann-broque) [inline srcs/Channel.cpp:35]: const

File: srcs/Channel.cpp [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+	userMap[client->getNickname()] = spClient;
+}
+
+const std::string Channel::getUserList() {
+	std::string userList;
+	for (std::map<std::string, SpecifiedClient>::iterator it = userMap.begin();
+		 it != userMap.end(); ++it) {
+		if (it == userMap.begin())
```
Comment (romann-broque) [inline srcs/Channel.cpp]: `
		if (it != userMap.begin())
                     userList += " ";
		userList += getSpecifiedNick(it->second);
`

File: srcs/Channel.cpp [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+			userList = getSpecifiedNick(it->second);
+		else
+			userList += " " + getSpecifiedNick(it->second);
+	}
+	return (userList);
+}
+
+void Channel::removeUser(Client *const client) {
```
Comment (romann-broque) [inline srcs/Channel.cpp]: const Client *const client

File: srcs/Client.cpp [hunk 19]
```diff
@@ -3,21 +3,30 @@
...
+/*   Updated: 2023/12/04 15:16:20 by mat              ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
 #include "Client.hpp"
 
-// Constructors
+#include <string.h>
+
+#include <iostream>
```
Comment (romann-broque) [inline srcs/Client.cpp:15]: remove iostream & string.h

File: srcs/Client.cpp [hunk 20]
```diff
@@ -3,21 +3,30 @@
...
 
-Client::Client() : _sockfd(0), _loginMask(EMPTY_LOGIN) {}
+// Constructors
 
 Client::Client(const int sockfd) : _sockfd(sockfd), _loginMask(EMPTY_LOGIN) {}
 
+Client::Client(const Client &ref)
```
Comment (romann-broque) [inline srcs/Client.cpp]: remove the copy constructor

File: srcs/ServerCmds.cpp [hunk 21]
```diff
@@ -34,6 +34,18 @@ static bool isNicknameValid(const std::string &nickname) {
 	return isdigit(nickname[0]) == false && areOnlyAuthorizedChar(nickname);
 }
 
+static std::string getFullMessage(const std::vector<std::string> &cmd) {
+	const size_t size = cmd.size();
+	std::string	 fullMessage;
+
+	if (size > 2) {
+		fullMessage = cmd[2];
+		size_t i = 3;
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: #define MESSAGE_START_INDEX 3

File: srcs/ServerCmds.cpp [hunk 22]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
 	sendFormattedMessage(PONG_MESSAGE, client);
 }
 
+void Server::sendJoinMessage(
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: `void Server::sendJoinMessage(
	const Channel *const channel, Client *const client, std::string channelName) {`

don't forget :
`	for (std::map<std::string, SpecifiedClient>::const_iterator it =
			 channel->userMap.begin();
		 it != channel->userMap.end(); ++it) {`

File: srcs/ServerCmds.cpp [hunk 23]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+	Channel *channel, Client *client, std::string channelName) {
+	std::string channelUserList;
+
+	printLog(client->getNickname());
+	sendPrivateMessage(JOIN_MESSAGE + channelName, client, client);
+	for (std::map<std::string, SpecifiedClient>::iterator it =
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: move into a Channel method -> sentToAll

File: srcs/ServerCmds.cpp [hunk 24]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
 	sendFormattedMessage(PONG_MESSAGE, client);
 }
 
+void Server::sendJoinMessage(
+	Channel *channel, Client *client, std::string channelName) {
+	std::string channelUserList;
+
+	printLog(client->getNickname());
+	sendPrivateMessage(JOIN_MESSAGE + channelName, client, client);
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: remove

File: srcs/ServerCmds.cpp [hunk 25]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+	for (std::map<std::string, SpecifiedClient>::iterator it =
+			 channel->userMap.begin();
+		 it != channel->userMap.end(); ++it) {
+		if (it->second.client->getSocketFd() != client->getSocketFd())
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: remove

File: srcs/ServerCmds.cpp [hunk 26]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+				JOIN_MESSAGE + channelName, client, it->second.client);
+	}
+	if (!channel->topic.empty())
+		sendFormattedMessage(TOPIC_JOIN_MESSAGE + channel->topic, client);
+	channelUserList = channel->getUserList();
+	sendFormattedMessage(UL_JOIN_MESSAGE + channelUserList, client);
+	sendFormattedMessage(EUL_JOIN_MESSAGE, client);
+}
+
+void Server::join(const std::vector<std::string> &cmd, Client *const client) {
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp:124]: `void Server::join(const std::vector<std::string> &cmd, const Client *const client) {`

File: srcs/ServerCmds.cpp [hunk 27]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+	if (!cmd[1].empty()) {
+		std::map<std::string, Channel *>::iterator it =
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: checker si possible

`const std::map<std::string, Channel *>::iterator it =...`

File: srcs/ServerCmds.cpp [hunk 28]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+			_channels.find("#" + cmd[1]);
+		if (it == _channels.end()) {
+			Channel *channel = new Channel(cmd[1], client);
+			_channels["#" + cmd[1]] = channel;
+			printLog("New Channel !");
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: remove logs

File: srcs/ServerCmds.cpp [hunk 29]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+			sendJoinMessage(channel, client, cmd[1]);
+		} else {
+			it->second->addNewUser(client);
+			printLog("Join Channel !");
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: remove logs

File: srcs/ServerCmds.cpp [hunk 30]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+	sendFormattedMessage(EUL_JOIN_MESSAGE, client);
+}
+
+void Server::join(const std::vector<std::string> &cmd, Client *const client) {
+	if (!cmd[1].empty()) {
+		std::map<std::string, Channel *>::iterator it =
+			_channels.find("#" + cmd[1]);
+		if (it == _channels.end()) {
+			Channel *channel = new Channel(cmd[1], client);
+			_channels["#" + cmd[1]] = channel;
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: define CHANNEL_PREFIX "#"

File: srcs/ServerCmds.cpp [hunk 31]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+			sendJoinMessage(it->second, client, cmd[1]);
+		}
+	}
+}
+
+void Server::privmsg(
+	const std::vector<std::string> &cmd, Client *const client) {
+	std::string fullMessage = getFullMessage(cmd);
+
+	printLog("1-> Searching for :" + cmd[1] + " channel");
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: remove logs

File: srcs/ServerCmds.cpp [hunk 32]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+	if (_channels.find("#" + cmd[1]) != _channels.end()) {
+		Channel *channel = _channels.find("#" + cmd[1])->second;
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: CHANNEL_PREFIX

File: srcs/ServerCmds.cpp [hunk 33]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+		printLog("2 -> Found a channel :" + channel->getName());
+		for (std::map<std::string, SpecifiedClient>::iterator it =
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: use sendToAll

File: srcs/ServerCmds.cpp [hunk 34]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+	} else if (_clientMap.getClient(cmd[1]) != NULL)
+		sendPrivateMessage(PRIVMSG_PREFIX + cmd[1] + " " + fullMessage, client,
+			_clientMap.getClient(cmd[1]));
+	else
+		printLog("Cound not send message");
+}
+
+void Server::part(const std::vector<std::string> &cmd, Client *const client) {
+	std::cout << cmd[1] << std::endl;
+	if (_channels.find("#" + cmd[1]) != _channels.end()) {
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: use sendToAll

Comment (romann-broque) [inline srcs/ServerCmds.cpp]: use CHANNEL_PREFIX

File: srcs/ServerCmds.cpp [hunk 35]
```diff
@@ -94,9 +106,86 @@ void Server::ping(const std::vector<std::string> &cmd, Client *const client) {
...
+			}
+		}
+	} else if (_clientMap.getClient(cmd[1]) != NULL)
+		sendPrivateMessage(PRIVMSG_PREFIX + cmd[1] + " " + fullMessage, client,
+			_clientMap.getClient(cmd[1]));
+	else
+		printLog("Cound not send message");
+}
+
+void Server::part(const std::vector<std::string> &cmd, Client *const client) {
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp:155]: const Client *const client

**Review by romann-broque — COMMENTED** [review]
good luck my friend

**Review by HelloThere-ImMat — COMMENTED** [review]
Finished Correction

File: srcs/ServerCmds.cpp (hunk 35 above)
Comment (HelloThere-ImMat) [reply srcs/ServerCmds.cpp:155]: Impossible, la commande user le modifie

**Review by HelloThere-ImMat — COMMENTED** [review]
All good

File: srcs/ServerCmds.cpp [hunk 36]
```diff
@@ -35,13 +35,15 @@ static bool isNicknameValid(const std::string &nickname) {
 }
 
 static std::string getFullMessage(const std::vector<std::string> &cmd) {
-	const size_t size = cmd.size();
-	std::string	 fullMessage;
-
+	std::string fullMessage;
+	int			size = cmd.size();
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: `	const size_t    size = cmd.size();
	size_t		i = PRIVMSG_START_INDEX;`

File: srcs/SendCmd.cpp [hunk 37]
```diff
@@ -0,0 +1,78 @@
...
+/*   Created: 2023/12/05 09:55:59 by mat               #+#    #+#             */
+/*   Updated: 2023/12/05 10:54:47 by mat              ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "SendCmd.hpp"
+
+// Static Funcs
+
+static std::string replacePatterns(std::string original,
```
Comment (romann-broque) [inline srcs/SendCmd.cpp]: `static std::string replacePatterns(const std::string &original,
	const std::string &pattern, const std::string &replacement)`

File: srcs/Channel.cpp [hunk 38]
```diff
@@ -3,56 +3,78 @@
...
+void Channel::removeUser(const Client *const client) {
+	std::map<std::string, SpecifiedClient>::iterator it;
 	std::string nickname = client->getNickname();
 	userMap.erase(nickname);
 }
 
 const std::string &Channel::getName() const { return _name; }
+
+void Channel::sendToChannel(
+	const Client *const client, std::string message, bool sendToSelf) const {
```
Comment (romann-broque) [inline srcs/Channel.cpp]: plutot enlever le bool sendToSelf pour faire 2 wrappers?

File: dependencies/Channel.hpp [hunk 39]
```diff
@@ -3,37 +3,39 @@
...
-} SpecifiedClient;
+struct SpecifiedClient {
+	const Client *client;
+	bool		  isOp;
+};
 
 class Channel {
    public:
-	Channel(std::string name, Client *const client);
+	Channel(const std::string name, const Client *const client);
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: `Channel(const std::string &name, const Client *const client);`

File: dependencies/Channel.hpp [hunk 40]
```diff
@@ -3,37 +3,39 @@
...
-	void								   removeUser(Client *const client);
-	const std::string					   getUserList();
-	const std::string					  &getName() const;
-	std::map<std::string, SpecifiedClient> userMap;
-	std::string							   topic;
+	void			   addNewUser(const Client *const client);
+	void			   removeUser(const Client *const client);
+	const std::string  getUserList() const;
+	const std::string &getName() const;
+	void			   sendToChannel(
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: `void			   sendToChannel(
					  const Client *const client, std::string message, bool sendToSelf) const;`
a remplacer par 2 wrappers. Par exemple: sendToChannel (en solo) et sendToAll (pour tout le monde).

File: dependencies/Channel.hpp [hunk 41]
```diff
@@ -3,37 +3,39 @@
...
+/*   By: mat <mat@student.42.fr>                    +#+  +:+       +#+        */
 /*                                                +#+#+#+#+#+   +#+           */
-/*   Created: 2023/12/04 14:37:48 by rbroque           #+#    #+#             */
-/*   Updated: 2023/12/04 14:40:14 by rbroque          ###   ########.fr       */
+/*   Created: 2023/11/29 12:53:30 by mat               #+#    #+#             */
+/*   Updated: 2023/12/05 11:25:17 by mat              ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
 #include <algorithm>
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: check if required

File: dependencies/Channel.hpp [hunk 42]
```diff
@@ -3,37 +3,39 @@
...
-// DEBUGING
-#include <iostream>
+#define OP_PREFIX "@"
 
-typedef struct s_specifiedClient {
-	Client *client;
-	bool	isOp;
-} SpecifiedClient;
+struct SpecifiedClient {
+	const Client *client;
```
Comment (romann-broque) [inline dependencies/Channel.hpp:20]: const Client *const client;

File: dependencies/Server.hpp [hunk 43]
```diff
@@ -6,68 +6,56 @@
...
-
 // Message
 
+#define JOIN_PREFIX		   "JOIN :"
+#define PART_PREFIX		   "PART "
 #define UL_JOIN_MESSAGE	   "353 <nick> = <arg> :"
 #define EUL_JOIN_MESSAGE   "366 <client> <arg> :End of /NAMES list."
 #define TOPIC_JOIN_MESSAGE "332 <client> <arg> :default"
 #define PONG_MESSAGE	   "PONG <servername> :<nick>"
+#define PRIVMSG_PREFIX	   "PRIVMSG "
```
Comment (romann-broque) [inline dependencies/Server.hpp]: peut-etre mettre les PREFIX ensemble?

File: srcs/Channel.cpp [hunk 44]
```diff
@@ -3,56 +3,78 @@
...
-		return ("@" + spClient.client->getNickname());
+		return (OP_PREFIX + spClient.client->getNickname());
 	return (spClient.client->getNickname());
 }
 
 // Methods
 
-Channel::Channel(std::string name, Client *const client) {
-	SpecifiedClient spClient = {.client = client, .isOp = true};
+Channel::Channel(const std::string name, const Client *const client)
```
Comment (romann-broque) [inline srcs/Channel.cpp]: `Channel::Channel(const std::string &name, const Client *const client)`

File: srcs/Channel.cpp [hunk 45]
```diff
@@ -3,56 +3,78 @@
...
+		if (it != userMap.begin())
+			userList += " ";
+		userList = getSpecifiedNick(it->second);
 	}
 	return (userList);
 }
 
-void Channel::removeUser(Client *const client) {
+void Channel::removeUser(const Client *const client) {
+	std::map<std::string, SpecifiedClient>::iterator it;
```
Comment (romann-broque) [inline srcs/Channel.cpp]: to be removed

**Review by romann-broque — COMMENTED** [review]
good luck again

File: srcs/ServerCmds.cpp [hunk 46]
```diff
@@ -34,6 +34,20 @@ static bool isNicknameValid(const std::string &nickname) {
 	return isdigit(nickname[0]) == false && areOnlyAuthorizedChar(nickname);
 }
 
+static std::string getFullMessage(const std::vector<std::string> &cmd) {
+	std::string fullMessage;
+	int			size = cmd.size();
+	int			i = PRIVMSG_START_INDEX;
```
Comment (romann-broque) [inline srcs/ServerCmds.cpp]: `size_t			i = PRIVMSG_START_INDEX;`

File: srcs/SendCmd.cpp (hunk 37 above)
Comment (HelloThere-ImMat) [reply srcs/SendCmd.cpp]: Impossible de mettre original en const car on appelle sa methode .replace

File: dependencies/Channel.hpp (hunk 42 above)
Comment (HelloThere-ImMat) [reply dependencies/Channel.hpp:20]: Impossible car assignation par copie dans la map

File: srcs/Channel.cpp (hunk 45 above)
Comment (HelloThere-ImMat) [reply srcs/Channel.cpp:53]: po compris, on en a besoin

**Review by HelloThere-ImMat — COMMENTED** [review]
All good

File: dependencies/Channel.hpp [hunk 47]
```diff
@@ -0,0 +1,40 @@
...
+
+class Channel {
+   public:
+	Channel(const std::string &name, const Client *const client);
+	~Channel();
+	void			   addNewUser(const Client *const client);
+	void			   removeUser(const Client *const client);
+	const std::string  getUserList() const;
+	const std::string &getName() const;
+	void sendToOthers(const Client *const client, std::string message) const;
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: `void sendToOthers(const Client *const client, const std::string &message) const;`

File: dependencies/Channel.hpp [hunk 48]
```diff
@@ -0,0 +1,40 @@
...
+	void sendToAll(const Client *const client, std::string message) const;
```
Comment (romann-broque) [inline dependencies/Channel.hpp]: `void sendToAll(const Client *const client, const std::string &message) const;`

File: srcs/Channel.cpp [hunk 49]
```diff
@@ -0,0 +1,85 @@
...
+/*   Created: 2023/11/29 13:18:12 by mat               #+#    #+#             */
+/*   Updated: 2023/12/05 14:31:36 by mdorr            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "Channel.hpp"
+
+// Static
+
+std::string getSpecifiedNick(const SpecifiedClient &spClient) {
```
Comment (romann-broque) [inline srcs/Channel.cpp]: `static std::string getSpecifiedNick(const SpecifiedClient &spClient) {`

File: srcs/SendCmd.cpp [hunk 50]
```diff
@@ -0,0 +1,78 @@
...
+/*   Created: 2023/12/05 09:55:59 by mat               #+#    #+#             */
+/*   Updated: 2023/12/05 14:17:06 by mdorr            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "SendCmd.hpp"
+
+// Static Funcs
+
+static std::string replacePatterns(std::string &original,
```
Comment (romann-broque) [inline srcs/SendCmd.cpp]: `static std::string replacePatterns(const std::string &original,
	const std::string &pattern, const std::string &replacement) {`
Il faut bien mettre en const l'original et creer une autre variable "newStr" par exemple pour contenir la string returned.

**Review by romann-broque — COMMENTED** [review]
Last :D

File: dependencies/Channel.hpp (hunk 47 above)
Comment (HelloThere-ImMat) [reply dependencies/Channel.hpp]: Impossible, car je n'envoie pas une ref mais la concatenation d'un define et d'une ref

File: dependencies/Channel.hpp (hunk 48 above)
Comment (HelloThere-ImMat) [reply dependencies/Channel.hpp]: Impossible, car je n'envoie pas une ref mais la concatenation d'un define et d'une ref

File: dependencies/Channel.hpp (hunk 47 above)
Comment (romann-broque) [reply dependencies/Channel.hpp]: juste le const alors :)

**Review by romann-broque — APPROVED** [review]

**Pull request merged**